client/tailscale/tailscale.go

```diff
@@ -30,45 +30,90 @@ var I_Acknowledge_This_API_Is_Unstable = false
 
 // TODO: use url.PathEscape() for deviceID and tailnets when constructing requests.
 
-// DefaultURL is the default base URL used for API calls.
-const DefaultURL = "https://api.tailscale.com"
+const defaultAPIBase = "https://api.tailscale.com"
 
 // maxSize is the maximum read size (10MB) of responses from the server.
-const maxReadSize int64 = 10 * 1024 * 1024
+const maxReadSize = 10 << 20
 
-// Client is needed to make different API calls to the Tailscale server.
-// It holds all the necessary information so that it can be reused to make
-// multiple requests for the same user.
-// Unless overridden, "api.tailscale.com" is the default BaseURL.
+// Client makes API calls to the Tailscale control plane API server.
+//
+// Use NewClient to instantiate one. Exported fields should be set before
+// the client is used and not changed thereafter.
 type Client struct {
-	// Tailnet is the globally unique identifier for a Tailscale network, such
+	// tailnet is the globally unique identifier for a Tailscale network, such
 	// as "example.com" or "user@gmail.com".
-	Tailnet    string
-	APIKey     string
-	BaseURL    string
+	tailnet string
+	// auth is the authentication method to use for this client.
+	// nil means none, which generally won't work, but won't crash.
+	auth AuthMethod
+
+	// BaseURL optionally specifies an alternate API server to use.
+	// If empty, "https://api.tailscale.com" is used.
+	BaseURL string
+
+	// HTTPClient optionally specifies an alternate HTTP client to use.
+	// If nil, http.DefaultClient is used.
 	HTTPClient *http.Client
 }
 
-// New is a convenience method for instantiating a new Client.
+func (c *Client) httpClient() *http.Client {
+	if c.HTTPClient != nil {
+		return c.HTTPClient
+	}
+	return http.DefaultClient
+}
+
+func (c *Client) baseURL() string {
+	if c.BaseURL != "" {
+		return c.BaseURL
+	}
+	return defaultAPIBase
+}
+
+// AuthMethod is the interface for API authentication methods.
+//
+// Most users will use AuthKey.
+type AuthMethod interface {
+	modifyRequest(req *http.Request)
+}
+
+// AuthKey is an AuthMethod for NewClient that authenticates requests
+// using an authkey.
+type AuthKey string
```

Comment:
APIKey
AuthKey is used specifically for registration of nodes

Reply:
braino